src/workerd/api/headers.c++

```diff
@@ -0,0 +1,821 @@
+    // held by the key, so we need to preserve the original casing for display
+    // purposes.
+    this->name = kj::mv(name);
+  }
+}
+
+Headers::Header::Header(
+    HeaderKey key, kj::Maybe<jsg::ByteString> name, kj::Vector<jsg::ByteString> values)
+    : key(kj::mv(key)),
+      name(kj::mv(name)),
+      values(kj::mv(values)) {}
+
+Headers::Header::Header(HeaderKey key, jsg::ByteString name, jsg::ByteString value)
+    : key(kj::mv(key)),
+      values(1) {
+  values.add(kj::mv(value));
+  if (getKeyName() != name) {
+    this->name = kj::mv(name);
+  }
+}
+
+kj::StringPtr Headers::Header::Header::getKeyName() const {
+  KJ_SWITCH_ONEOF(key) {
+    KJ_CASE_ONEOF(commonId, uint) {
+      return COMMON_HEADER_NAMES[commonId];
+    }
+    KJ_CASE_ONEOF(uncommonKey, UncommonHeaderKey) {
+      return uncommonKey.getName();
+    }
+  }
+  KJ_UNREACHABLE;
+}
+
+kj::StringPtr Headers::Header::getHeaderName() const {
+  KJ_IF_SOME(preservedName, name) {
+    return preservedName;
+  }
+  return getKeyName();
+}
+
+Headers::Header Headers::Header::clone() const {
+  return Header(cloneHeaderKey(key),
+      name.map([](const kj::String& n) { return jsg::ByteString(kj::str(n)); }),
+      KJ_MAP(value, values) { return jsg::ByteString(kj::str(value)); });
+}
+
+bool Headers::HeaderCallbacks::matches(Header& header, const HeaderKey& other) {
+  return headerKeyEquals(header.key, other);
+}
+
+bool Headers::HeaderCallbacks::matches(Header& header, kj::StringPtr otherName) {
+  return matches(header, getHeaderKeyFor(otherName));
+}
+
+bool Headers::HeaderCallbacks::matches(Header& header, capnp::CommonHeaderName commondId) {
+  KJ_IF_SOME(headerCommonId, header.key.tryGet<uint>()) {
+    return headerCommonId == static_cast<uint>(commondId);
+  }
+  return false;
+}
+
+kj::uint Headers::HeaderCallbacks::hashCode(const HeaderKey& key) {
+  KJ_SWITCH_ONEOF(key) {
+    KJ_CASE_ONEOF(commonId, uint) {
+      return kj::hashCode(commonId);
+    }
+    KJ_CASE_ONEOF(uncommonKey, UncommonHeaderKey) {
+      return uncommonKey.hashCode();
+    }
+  }
+  KJ_UNREACHABLE;
+}
+
+kj::uint Headers::HeaderCallbacks::hashCode(capnp::CommonHeaderName commondId) {
+  return kj::hashCode(static_cast<uint>(commondId));
+}
+
+Headers::Headers(jsg::Lock& js, jsg::Dict<jsg::ByteString, jsg::ByteString> dict)
+    : guard(Guard::NONE) {
+  headers.reserve(dict.fields.size());
+  for (auto& field: dict.fields) {
+    append(js, kj::mv(field.name), kj::mv(field.value));
+  }
+}
+
+Headers::Headers(jsg::Lock& js, const Headers& other): guard(Guard::NONE) {
+  headers.reserve(other.headers.size());
+  for (auto& header: other.headers) {
+    // There really shouldn't be any duplicate headers in other, but just in case, use upsert
+    // and we'll just ignore duplicates.
+    headers.upsert(header.clone(), [](auto&, auto&&) {});
+  }
+}
+
+Headers::Headers(jsg::Lock& js, const kj::HttpHeaders& other, Guard guard): guard(Guard::NONE) {
```

Comment:
This constructor -- converting KJ headers to api::Headers -- can be much faster if it used the KJ header ID -> common header ID mapping available in `HttpOverCapnpFactory`. Then you don't need to look at string names at all for common headers.

See `HttpOverCapnpFactory::headersToCapnp()` for an example of how to implement, though that code specifically converts to capnp format, so you'll need different (but similarly-shaped) code here.